lib/hmmlearn/hmm.py

```diff
@@ -307,18 +309,246 @@ def _do_mstep(self, stats):
+                "The previous version was actually implenting CategoricalHMM. This new "
+                "implementation follows the standard definition for a Multinomial distribution, "
+                "e.g. as in https://en.wikipedia.org/wiki/Multinomial_distribution"
+                "See these issues for details: "
+                "\nhttps://github.com/hmmlearn/hmmlearn/issues/335 "
+                "\nhttps://github.com/hmmlearn/hmmlearn/issues/340 "
+                "\nThis is also why n_components and n_trials are now required arguments.")
+
+    # From the API: Return a mapping of fittable parameter names 
+    # (as in self.params) to the number of corresponding scalar parameters 
+    # that will actually be fitted. This is used to detect whether the user 
+    # did not pass enough data points for a non-degenerate fit.
+    def _get_n_fit_scalars_per_param(self):
+        nc = self.n_components
+        nf = self.n_features
+        return {
+            "s": nc - 1,
+            "t": nc * (nc - 1),
+            "e": nc * (nf - 1),
+        }
+
+    
+    def _check_and_set_multinomial_n_features(self, X):
+        """
+        Check if ``X`` is a sample from a multinomial distribution, i.e. an
+        array of non-negative integers, summing up to n_trials.
+        """
+        n_samples, n_features = X.shape
+        if n_features != self.n_features:
+            raise ValueError("Feature dimension should be (n_samples, n_features)")
+        if not (X.sum(axis=1) == self.n_trials).all():
+            raise ValueError("Total count for each sample should add up to number of trials")
+        if not np.issubdtype(X.dtype, np.integer):
+            raise ValueError("Symbols should be integers")
+        if X.min() < 0:
+            raise ValueError("Symbols should be nonnegative")
+        if hasattr(self, "n_features"):
+            if self.n_trials < X.max():
+                raise ValueError(
+                    " Largest number of trials in a sample is {} " 
+                    " but the model takes max of {} trials".format(X.max(), self.n_trials))
+        else:
+            self.n_features = X.max() + 1
+
+            
+    def _init(self, X):
+        self._check_and_set_multinomial_n_features(X)
+        super()._init(X)
+        self.random_state = check_random_state(self.random_state)
+        
+        if isinstance(self.transmat_prior, np.ndarray):
+            self.transmat_ = self.transmat_prior.copy()
+        else:
+            self.transmat_ = np.random.random((self.n_components, self.n_components))
+            normalize(self.transmat_, axis=1)
+        
+        if isinstance(self.startprob_prior, np.ndarray):
+            self.startprob_ = self.startprob_prior.copy()
+        else:
+            self.startprob_ = np.random.random(self.n_components)
+            normalize(self.startprob_)
+
+        if isinstance(self.emission_prior, np.ndarray):
+            self.emissionprob_ = self.emission_prior.copy()
```

Comment:
copying emission_prior to emissionprob seems wrong, as this is just duplicating the functionality of directly setting the emissionprob attribute, which remains the only approach that works for other hmms (this is also in fact a source of confusion for some users).  I guess I would drop emission_prior for now, or perhaps, if you want a way to specify priors, these should be understood as "pseudocounts" added to the normal counts? (as in https://github.com/hmmlearn/hmmlearn/issues/58#issuecomment-155572844)

Reply:
Thanks for the reference, I did indeed misunderstand the purpose of emission_prior.